com.unity.visualeffectgraph/Editor/Data/VFXDataParticle.cs

```diff
@@ -191,8 +192,188 @@ public BucketInfo[] GetBucketLayoutInfo()
+        {
+            [SerializeField]
+            private List<string> m_PropertyNames = new List<string>();
+
+            [SerializeField]
+            private List<float> m_PropertyValues = new List<float>();
+
+            public RenderStateProperties(Material material)
+            {
+                var shader = material.shader;
+
+                if (shader == null)
+                    return;
+
+                var properties = ShaderUtil.GetMaterialProperties(new Object[] {material});
+
+                foreach (var property in properties)
+                {
+                    if (property.type != MaterialProperty.PropType.Float)
+                        continue;
+
+                    m_PropertyNames.Add(property.name);
+                    m_PropertyValues.Add(property.floatValue);
+                }
+            }
+
+            public void ConfigureMaterial(Material material)
+            {
+                for (int i = 0; i < Mathf.Min(m_PropertyNames.Count, m_PropertyValues.Count); i++)
+                    material.SetFloat(m_PropertyNames[i], m_PropertyValues[i]);
+            }
+        }
+
+        [SerializeField]
+        private List<VFXContext> m_Contexts = new List<VFXContext>();
+
+        [SerializeField]
+        private List<RenderStateProperties> m_CachedRenderState = new List<RenderStateProperties>();
+
+        public void OnBeforeSerialize() => CacheRenderStates();
+        public void OnAfterDeserialize() => LoadRenderStates();
+
+        public void CacheRenderStates()
+        {
+            m_Contexts.Clear();
+            m_CachedRenderState.Clear();
+            foreach (var kvp in m_MaterialMap)
+            {
+                m_Contexts.Add(kvp.Key);
+                m_CachedRenderState.Add(new RenderStateProperties(kvp.Value));
+            }
+        }
+
+        public void LoadRenderStates()
+        {
+            m_RenderStatePropertyMap = new Dictionary<VFXContext, RenderStateProperties>();
+            for (int i = 0; i != Math.Min(m_Contexts.Count, m_CachedRenderState.Count); i++)
+                m_RenderStatePropertyMap.Add(m_Contexts[i], m_CachedRenderState[i]);
+        }
+    }
+
     class VFXDataParticle : VFXData, ISpaceable
     {
+        // Maintain a collection since one particle data may contain multiple contexts with multiple materials.
+        [SerializeField]
+        private VFXMaterialCollection m_MaterialCollection = new VFXMaterialCollection();
+
+        public Material GetOrCreateMaterial(VFXContext context) => m_MaterialCollection.GetOrCreate(context);
+
+        public void SyncContextMaterial(VFXContext context) => m_MaterialCollection.TrySyncRenderStateProperties(context);
+
+        static VFXMapping InjectMaterialState() => new VFXMapping("injectMaterialState", 1);
+
+        static void AppendMaterialParameters(Material material, List<VFXMapping> parameters, VFXContextCompiledData contextData, VFXExpressionGraph expressionGraph)
```

Comment:
This will also be removed after the C++ PR lands. It is a temporary hack to unblock development/testing that uses the VFXMappings to inject the state of the material into the underlying C++ material created by VFX.